kube-core/src/schema.rs

```diff
@@ -0,0 +1,72 @@
+//! Utilities for managing [`CustomResourceDefinition`] schemas
+//!
+//! [`CustomResourceDefinition`]: `k8s_openapi::apiextensions_apiserver::pkg::apis::apiextensions::v1::CustomResourceDefinition`
+
+use std::collections::btree_map::Entry;
+
+use schemars::{
+    schema::{ObjectValidation, Schema, SchemaObject},
+    visit::Visitor,
+};
+
+/// schemars [`Visitor`] that rewrites a [`Schema`] to conform to Kubernetes' "structural schema" rules
+///
+/// This is not intended for direct use, but is used by `kube::derive`'s `#[derive(CustomResource)]`.
+#[derive(Debug, Clone)]
+pub struct StructuralSchemaRewriter;
+
+impl Visitor for StructuralSchemaRewriter {
+    fn visit_schema_object(&mut self, schema: &mut schemars::schema::SchemaObject) {
+        schemars::visit::visit_schema_object(self, schema);
+        if let Some(one_of) = schema
+            .subschemas
+            .as_mut()
+            .and_then(|subschemas| subschemas.one_of.as_mut())
+        {
+            let common_obj = schema
+                .object
+                .get_or_insert_with(|| Box::new(ObjectValidation::default()));
+            for variant in one_of {
+                if let Schema::Object(SchemaObject {
+                    instance_type: variant_type,
+                    object: Some(variant_obj),
+                    ..
+                }) = variant
+                {
+                    // Move all properties
+                    let variant_properties = std::mem::take(&mut variant_obj.properties);
+                    for (property_name, property) in variant_properties {
+                        match common_obj.properties.entry(property_name) {
+                            Entry::Occupied(entry) => panic!(
+                                "property {:?} is already defined in another enum variant",
```

Comment:
In practice this means that trying to use the same tag for multiple variants will fail, but I'm not sure whether Serde will accept that anyway. For example:

```rust
enum Foo {
    Spam {
        expiry_date: String,
    },
    #[serde(rename = "Spam")]
    Eggs {
        expiry_date: Int,
    },
}
```

You might also have issues with internal tagging (when defining `#[serde(tag = "type")]`, causing serde to generate `{"type": "Spam", "expiry_date": "foo"}` rather than `{"Spam": {"expiry_date": "foo"}}`), as well as untagged enums (where no tag is emitted as all, and Serde just tries the variants in order). I don't think that's a massive issue, since external tagging is generally the convention in K8s anyway.

Reply:
Ah, in which case, I think it would be useful to write this under `kube-derive/src/lib.rs` under the `## Advanced Feautures`
section. something like:

```markdown
### Using enums

When enums are present, `kube-derive` will rewrite the schemas for these using a [`schemars::Visitors`](https://docs.rs/schemars/latest/schemars/visit/trait.Visitor.html).
This ensures that the generated [schema is structural](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema) and accepted by Kubernetes.

Note that not all types of [enums representations](https://serde.rs/enum-representations.html) are guaranteed to work:

- untagged enums can produce conflicting variant types
- internally tagged enums can produce duplicate properties on one enum variant

for best results use the Kubernetes convention of using [externally tagged enums](https://serde.rs/enum-representations.html#externally-tagged).
```

maybe we should also export these as an error as well so that kube-derive will fail witha  nice message?

Reply:
1. That seems like a good idea, but I think the visitor stuff is a bit of an implementation detail. Maybe it's enough to just write down (the equivalent of) "here's what we support, here's what we don't" while pointing to the structural stuff for reasoning?

Reply:
2. or third way - long term - ask schemars to change the `Visitor` setup - but not sure that's something they will want.

Reply:
08974ccf549274c73d04fcaa019104f95e2116e7